lib/ttfunk/table/os2.rb

```diff
@@ -256,6 +265,80 @@ def group_original_code_points_by_bit(os2)
             end
           end
         end
+
+        def avg_char_width_for(os2, subset)
+          if subset.microsoft_symbol?
+            avg_ms_symbol_char_width_for(os2, subset)
+          else
+            avg_weighted_char_width_for(os2, subset)
+          end
+        end
+
+        def avg_ms_symbol_char_width_for(os2, subset)
+          total_width = 0
+          num_glyphs = 0
+
+          # use new -> old glyph mapping in order to include compound glyphs
+          # in the calculation
+          subset.new2old_glyph.each do |_, old_gid|
+            if (metric = os2.file.horizontal_metrics.for(old_gid))
+              total_width += metric.advance_width
+              num_glyphs += 1 if metric.advance_width > 0
+            end
+          end
+
+          return 0 if num_glyphs == 0
+
+          total_width / num_glyphs # this should be a whole number
+        end
+
+        def avg_weighted_char_width_for(os2, subset)
+          # make sure the subset includes the space char
+          return 0 unless subset.to_unicode_map[CODEPOINT_SPACE]
+
+          space_gid = os2.file.cmap.unicode.first[CODEPOINT_SPACE]
+          space_hm = os2.file.horizontal_metrics.for(space_gid)
+          return 0 unless space_hm
```

Comment:
I don't quite follow. Why do we assume _average subset char width_ 0 if there's no space in the subset?

Intuitively it doesn't feel right.

Reply:
You're absolutely right, nice catch. As I recall, one of the font specs (maybe CFF?) requires the space character (0x32) to be included in all fonts. I adapted most of this code from Font-Validator, which raises an exception in this case (`FastMapUnicodeToGlyphID` will return `0xFFFFFFFF` which will be larger than `GetMaxpNumGlyphs`, causing an exception). I suppose that's what we should do too.